File: geompp_tests/src/test_brect2d.cpp
#include "brect2d.hpp"
#include "point2d.hpp"
#include "utils.hpp"
#include "vector2d.hpp"

#include <gtest/gtest.h>

#include <cmath>

namespace g = geompp;

namespace geompp_tests {

class BRect2DTest : public ::testing::Test {
 protected:
  void SetUp() override { g::DECIMAL_PRECISION = g::DP_THREE; }
  void TearDown() override { g::DECIMAL_PRECISION = g::DP_THREE; }
};

// ── Constructor ──────────────────────────────────────────────────────────────

TEST_F(BRect2DTest, ConstructorEmpty_Throws) {
  ASSERT_THROW(g::BRect2D(std::vector<g::Point2D>{}), std::invalid_argument);
}

TEST_F(BRect2DTest, ConstructorSinglePoint_Throws) {
  ASSERT_THROW(g::BRect2D(std::vector<g::Point2D>{{3.0, 4.0}}), std::invalid_argument);
}

TEST_F(BRect2DTest, ConstructorTwoPoints_Throws) {
  ASSERT_THROW(g::BRect2D(std::vector<g::Point2D>{{0.0, 0.0}, {4.0, 0.0}}), std::invalid_argument);
}

TEST_F(BRect2DTest, ConstructorAxisAlignedSquare) {
  // unit square; any axis-aligned OBB has area = 1
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {1.0, 0.0}, {1.0, 1.0}, {0.0, 1.0}};
  auto r = g::BRect2D(pts);
  ASSERT_TRUE(r.center().AlmostEquals(g::Point2D(0.5, 0.5)));
  ASSERT_NEAR(r.half_len_u(), 0.5, 1e-9);
  ASSERT_NEAR(r.half_len_v(), 0.5, 1e-9);
  ASSERT_NEAR(r.area(), 1.0, 1e-9);
}

TEST_F(BRect2DTest, ConstructorAxisAlignedRectangle) {
  // 4x2 rectangle: OBB aligned with x-axis (first hull edge wins)
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r = g::BRect2D(pts);
  ASSERT_TRUE(r.center().AlmostEquals(g::Point2D(2.0, 1.0)));
  ASSERT_NEAR(r.area(), 8.0, 1e-9);
  ASSERT_NEAR(r.width(), 4.0, 1e-9);
  ASSERT_NEAR(r.height(), 2.0, 1e-9);
}

TEST_F(BRect2DTest, ConstructorNonConvex_SmallArea) {
  // L-shaped non-convex polygon — OBB is tighter than AABB
  std::vector<g::Point2D> pts = {
    {0.0, 0.0}, {4.0, 0.0}, {4.0, 1.0},
    {1.0, 1.0}, {1.0, 4.0}, {0.0, 4.0},
  };
  auto r = g::BRect2D(pts);
  // AABB would be 4x4 = 16; OBB should cover 4x4 area (same in this case)
  ASSERT_TRUE(r.Contains(g::Point2D(0.5, 0.5)));
  ASSERT_TRUE(r.Contains(g::Point2D(3.5, 0.5)));
}

TEST_F(BRect2DTest, ConstructorAllPointsContained) {
  std::vector<g::Point2D> pts = {
    {0.0, 0.0}, {3.0, 0.0}, {3.0, 2.0}, {0.0, 2.0},
    {1.0, 0.5}, {2.0, 1.5},  // interior points
  };
  auto r = g::BRect2D(pts);
  for (auto const& p : pts)
    ASSERT_TRUE(r.Contains(p)) << "OBB must contain all input points";
}

// ── Accessors ────────────────────────────────────────────────────────────────

TEST_F(BRect2DTest, Accessors_AxisesAreUnitVectors) {
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {2.0, 0.0}, {2.0, 1.0}, {0.0, 1.0}};
  auto r = g::BRect2D(pts);
  ASSERT_NEAR(r.axis_u().Length(), 1.0, 1e-9);
  ASSERT_NEAR(r.axis_v().Length(), 1.0, 1e-9);
}

TEST_F(BRect2DTest, Accessors_AxesOrthogonal) {
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {3.0, 1.0}, {2.0, 3.0}, {-1.0, 2.0}};
  auto r = g::BRect2D(pts);
  double dot = r.axis_u().Dot(r.axis_v());
  ASSERT_NEAR(dot, 0.0, 1e-9);
}

TEST_F(BRect2DTest, Accessors_WidthHeightArea) {
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {6.0, 0.0}, {6.0, 3.0}, {0.0, 3.0}};
  auto r = g::BRect2D(pts);
  ASSERT_NEAR(r.width(), 6.0, 1e-9);
  ASSERT_NEAR(r.height(), 3.0, 1e-9);
  ASSERT_NEAR(r.area(), 18.0, 1e-9);
}

// ── Corners ──────────────────────────────────────────────────────────────────

TEST_F(BRect2DTest, Corners_FourDistinctPoints) {
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r = g::BRect2D(pts);
  auto corners = r.Corners();
  ASSERT_EQ(corners.size(), 4u);
  // all four corners must be contained in the rectangle
  for (auto const& c : corners)
    ASSERT_TRUE(r.Contains(c));
}

// ── Contains ─────────────────────────────────────────────────────────────────

TEST_F(BRect2DTest, Contains_Center_True) {
  geompp::DECIMAL_PRECISION = 6;
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r = g::BRect2D(pts);
  ASSERT_TRUE(r.Contains(g::Point2D(2.0, 1.0)));   // center
}

TEST_F(BRect2DTest, Contains_Interior_True) {
  geompp::DECIMAL_PRECISION = 6;
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r = g::BRect2D(pts);
  ASSERT_TRUE(r.Contains(g::Point2D(1.0, 0.5)));
  ASSERT_TRUE(r.Contains(g::Point2D(3.5, 1.8)));
}

TEST_F(BRect2DTest, Contains_Boundary_True) {
  geompp::DECIMAL_PRECISION = 6;
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r = g::BRect2D(pts);
  ASSERT_TRUE(r.Contains(g::Point2D(0.0, 0.0)));   // corner
  ASSERT_TRUE(r.Contains(g::Point2D(4.0, 1.0)));   // edge midpoint
}

TEST_F(BRect2DTest, Contains_Outside_False) {
  geompp::DECIMAL_PRECISION = 6;
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r = g::BRect2D(pts);
  ASSERT_FALSE(r.Contains(g::Point2D(5.0, 1.0)));   // beyond u-axis
  ASSERT_FALSE(r.Contains(g::Point2D(2.0, 3.0)));   // beyond v-axis
  ASSERT_FALSE(r.Contains(g::Point2D(-1.0, 1.0)));  // negative u
}

// ── AlmostEquals / operator== ─────────────────────────────────────────────────

TEST_F(BRect2DTest, AlmostEquals_SameRect) {
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r1 = g::BRect2D(pts);
  auto r2 = g::BRect2D(pts);
  ASSERT_TRUE(r1.AlmostEquals(r2));
  ASSERT_EQ(r1, r2);
}

TEST_F(BRect2DTest, AlmostEquals_DifferentRect) {
  std::vector<g::Point2D> pts1 = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  std::vector<g::Point2D> pts2 = {{0.0, 0.0}, {6.0, 0.0}, {6.0, 2.0}, {0.0, 2.0}};
  auto r1 = g::BRect2D(pts1);
  auto r2 = g::BRect2D(pts2);
  ASSERT_FALSE(r1.AlmostEquals(r2));
  ASSERT_NE(r1, r2);
}

// ── Copy / Assignment ─────────────────────────────────────────────────────────

TEST_F(BRect2DTest, CopyConstructor) {
  std::vector<g::Point2D> pts = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  auto r1 = g::BRect2D(pts);
  auto r2 = r1;
  ASSERT_TRUE(r1.AlmostEquals(r2));
}

TEST_F(BRect2DTest, Assignment) {
  std::vector<g::Point2D> pts1 = {{0.0, 0.0}, {4.0, 0.0}, {4.0, 2.0}, {0.0, 2.0}};
  std::vector<g::Point2D> pts2 = {{0.0, 0.0}, {1.0, 0.0}, {1.0, 1.0}, {0.0, 1.0}};
  auto r1 = g::BRect2D(pts1);
  auto r2 = g::BRect2D(pts2);
  r2 = r1;
  ASSERT_TRUE(r2.AlmostEquals(r1));
  r1 = r1;  // self-assignment
  ASSERT_TRUE(r1.AlmostEquals(g::BRect2D(pts1)));
}

}  // namespace geompp_tests

File: geompp_python/tests/test_brect2d.py
"""
brect2d binding tests.
"""

import math
import os
import tempfile
import pytest
import geompp

from ._helpers import approx


class TestBRect2D:
    def test_empty_throws(self):
        with pytest.raises(Exception):
            geompp.BRect2D([])

    def test_single_point_throws(self):
        with pytest.raises(Exception):
            geompp.BRect2D([geompp.Point2D(3, 4)])

    def test_two_points_throws(self):
        with pytest.raises(Exception):
            geompp.BRect2D([geompp.Point2D(0, 0), geompp.Point2D(4, 0)])

    def test_axis_aligned_rectangle_center(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
               geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        r = geompp.BRect2D(pts)
        assert r.center.almost_equals(geompp.Point2D(2, 1))
        assert approx(r.area, 8.0)
        assert approx(r.width, 4.0)
        assert approx(r.height, 2.0)

    def test_all_points_contained(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(3, 0),
               geompp.Point2D(3, 2), geompp.Point2D(0, 2),
               geompp.Point2D(1, 0.5), geompp.Point2D(2, 1.5)]
        r = geompp.BRect2D(pts)
        for p in pts:
            assert r.contains(p), f"OBB must contain input point {p}"

    def test_axes_are_unit_vectors(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(2, 0),
               geompp.Point2D(2, 1), geompp.Point2D(0, 1)]
        r = geompp.BRect2D(pts)
        assert approx(r.axis_u.length(), 1.0)
        assert approx(r.axis_v.length(), 1.0)

    def test_axes_orthogonal(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(3, 0),
               geompp.Point2D(3, 2), geompp.Point2D(0, 2)]
        r = geompp.BRect2D(pts)
        assert approx(r.axis_u.dot(r.axis_v), 0.0)

    def test_corners_returns_four_points(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
               geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        r = geompp.BRect2D(pts)
        corners = r.corners()
        assert len(corners) == 4
        for c in corners:
            assert r.contains(c)

    def test_contains_center(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
               geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        r = geompp.BRect2D(pts)
        assert r.contains(geompp.Point2D(2, 1))

    def test_contains_boundary(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
               geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        r = geompp.BRect2D(pts)
        assert r.contains(geompp.Point2D(0, 0))    # corner
        assert r.contains(geompp.Point2D(4, 1))    # edge midpoint

    def test_contains_outside(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
               geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        r = geompp.BRect2D(pts)
        assert not r.contains(geompp.Point2D(5, 1))
        assert not r.contains(geompp.Point2D(2, 3))

    def test_almost_equals(self):
        pts = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
               geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        r1 = geompp.BRect2D(pts)
        r2 = geompp.BRect2D(pts)
        assert r1.almost_equals(r2)
        assert r1 == r2

    def test_not_almost_equals_different(self):
        pts1 = [geompp.Point2D(0, 0), geompp.Point2D(4, 0),
                geompp.Point2D(4, 2), geompp.Point2D(0, 2)]
        pts2 = [geompp.Point2D(0, 0), geompp.Point2D(6, 0),
                geompp.Point2D(6, 2), geompp.Point2D(0, 2)]
        r1 = geompp.BRect2D(pts1)
        r2 = geompp.BRect2D(pts2)
        assert not r1.almost_equals(r2)
